iep-007/README.adoc

```diff
@@ -0,0 +1,153 @@
+
+
+== Specification
+
+Currently there is two main strategies to upgrade a cluster. +
+
+. Upgrading an existing cluster.
+. Migration on a second cluster.
+
+As Azure do not (yet) provide any tools to upgrade existing clusters, we have to upgrade them manually.  +
+It appears to be easier and safer to deploy a new cluster and re-deploy all resources on the new one.
+
+Until the new cluster stays in the same region than the previous one, we can keep persistent data from blob storage and attach them to the new cluster. +
+The only important element that we loose when we migrate to a new cluster, is the cluster public IP. +
+This means that we need to update 'nginx.azure.jenkins.io' with the public IP.
+
+=== Migration Process
+
+IMPORTANT: The old cluster must be kept until the new one is ready to served requests.
+
+==== Step 1 : Backup
+
+. Backup secrets containing Letsencrypt certificates. (Manual operation)
+----
+mkdir ~/tls
+.bin/kubectl get secret accountapp-tls --export=true  -o yaml > ~/tls/accountapp.yaml
+.bin/kubectl get secret pluginsite-tls --export=true  -o yaml > ~/tls/pluginsite.yaml
+.bin/kubectl get secret repo-proxy-tls --export=true  -o yaml > ~/tls/repo-proxy.yaml
+----
+
+==== Step 2 : Deploy the new cluster
+
+. Add a second k8s resource in github.com/jenkins-infra/azure, named 'peak8s' (Require PR on jenkins-infra/azure)
+. Update following hieraconfig variables with new the k8s cluster informations(Require PR on jenkins-infra/jenkins-infra)
+----
+profile::kubernetes::kubectl::server 
+profile::kubernetes::kubectl::username 
+profile::kubernetes::kubectl::clustername 
+profile::kubernetes::kubectl::certificate_authority_data
+profile::kubernetes::kubectl::client_certificate_data
+profile::kubernetes::kubectl::client_key_data
+----
+
+. Remove /home/k8s/resources (Manual operation)
+. Re-apply puppet agent with updated variables (Manual operation)
+. Wait for the new public IP (Manual operation)
+----
+kubectl get service nginx --namespace nginx-ingress
+----
+. Restore backed up secrets containing Letsencrypt certificates on the new cluster (Manual operation)
+----
+.bin/kubectl apply -f ~/tls/accountapp.yaml
+.bin/kubectl apply -f ~/tls/pluginsite.yaml
+.bin/kubectl apply -f ~/tls/repo-proxy.yaml
+----
+. Validate that everything work as expected (Manual operation)
+----
+curl --header 'Host: plugins.jenkins.io' 'https://<new_public_ip>
+curl --header 'Host: repo-proxy.jenkins.io' 'https://<new_public_ip>
+curl --header 'Host: accounts.jenkins.io' 'https://<new_public_ip>
+----
+
+==== Step 3: Update DNS from old to new cluster
+. Update nginx.azure.jenkins.io with the new cluster public IP (Require PR on jenkins-infra/jenkins-infra)
+
+[NOTE]
+During DNS update, requests will be send either to the new cluster, either to the old cluster.
+Users shouldn't detect any differences.
+
+==== Step 4: Remove the old cluster
+. Remove k8s.tf from jenkins-infra/azure (Require PR on jenkins-infra/azure)
```

Comment:
I think that's reasonable, less chance of somebody accidentally deleting production infrastructure